go/vrf/conversion.go

```diff
@@ -79,34 +78,30 @@ func secg1EncodeCompressed(curve elliptic.Curve, x, y *big.Int) []byte {
 
 var errInvalidPoint = errors.New("invalid point")
 
-func secg1Decode(curve elliptic.Curve, data []byte) (x, y *big.Int, err error) {
+func unmarshalCompressed(curve elliptic.Curve, data []byte) (x, y *big.Int, err error) {
 	byteLen := (curve.Params().BitSize + 7) >> 3
-	if (data[0] &^ 1) != 2 {
+	if (data[0] &^ 1) != 2 { // compressed form
 		return nil, nil, errors.New("unrecognized point encoding")
 	}
 	if len(data) != 1+byteLen {
 		return nil, nil, errors.New("invalid length for curve")
 	}
 
 	// Based on Routine 2.2.4 in NIST Mathematical routines paper
-	params := curve.Params()
-	tx := new(big.Int).SetBytes(data[1 : 1+byteLen])
-	y2 := y2(params, tx)
-	sqrt := defaultSqrt
-	ty := sqrt(y2, params.P)
-	if ty == nil {
+	x = new(big.Int).SetBytes(data[1:])
+	y2 := y2(curve.Params(), x)
+	y = new(big.Int).ModSqrt(y2, curve.Params().P)
+	if y == nil {
 		return nil, nil, errInvalidPoint // "y^2" is not a square
 	}
-	var y2c big.Int
-	y2c.Mul(ty, ty).Mod(&y2c, params.P)
-	if y2c.Cmp(y2) != 0 {
+	if !curve.IsOnCurve(x, y) {
 		return nil, nil, errInvalidPoint // sqrt(y2)^2 != y2: invalid point
```

Comment:
This comment is out of date. It should be something like "(x, y) is not a valid curve point"

Reply:
Also, I'm not sure that decompression errors like this are possible. What I mean is: because you've received a compressed point and you calculated the corresponding y value yourself, it should always be true that curve.IsOnCure(x, y) is true.

Reply:
Removed comment as code is more self documenting now